In [18]:
import requests
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import json

import warnings
warnings.filterwarnings("ignore")

In [56]:
q = []
for i in range(10807, 18813):
    q.append(i)
q = sorted(q)
len(q)

8006

In my project total 8006 reviews were there till 2023, so 8006 api calls

In [60]:
q[:1000]

[10807,
 10808,
 10809,
 10810,
 10811,
 10812,
 10813,
 10814,
 10815,
 10816,
 10817,
 10818,
 10819,
 10820,
 10821,
 10822,
 10823,
 10824,
 10825,
 10826,
 10827,
 10828,
 10829,
 10830,
 10831,
 10832,
 10833,
 10834,
 10835,
 10836,
 10837,
 10838,
 10839,
 10840,
 10841,
 10842,
 10843,
 10844,
 10845,
 10846,
 10847,
 10848,
 10849,
 10850,
 10851,
 10852,
 10853,
 10854,
 10855,
 10856,
 10857,
 10858,
 10859,
 10860,
 10861,
 10862,
 10863,
 10864,
 10865,
 10866,
 10867,
 10868,
 10869,
 10870,
 10871,
 10872,
 10873,
 10874,
 10875,
 10876,
 10877,
 10878,
 10879,
 10880,
 10881,
 10882,
 10883,
 10884,
 10885,
 10886,
 10887,
 10888,
 10889,
 10890,
 10891,
 10892,
 10893,
 10894,
 10895,
 10896,
 10897,
 10898,
 10899,
 10900,
 10901,
 10902,
 10903,
 10904,
 10905,
 10906,
 10907,
 10908,
 10909,
 10910,
 10911,
 10912,
 10913,
 10914,
 10915,
 10916,
 10917,
 10918,
 10919,
 10920,
 10921,
 10922,
 10923,
 10924,
 10925,
 10926,
 10927,
 10928,
 10929,
 10930,
 10931,


In [68]:
data_20_1 = data_20
data_21_1 = data_21
data_22_1 = data_22
data_23_1 = data_23

### Main code

In [64]:
url = "http://{api-service-name}/{api-endpoints}/{value-}"                             #API endpoint
c = 0 
for i in q[:8007]:
    u = url + str(i) + "/details"
    dat = requests.get(u, headers={"Accept": "application/json"}).json()
    if ('code' not in dat) and (dat['state'] == 'Closed'):
        if(dat['closeDate'] >= "2020-01-01" and dat['closeDate'] <= "2020-12-31"):     #Grouping reviews by year
            data_20.append(dat)
        if(dat['closeDate'] >= "2021-01-01" and dat['closeDate'] <= "2021-12-31"):
            data_21.append(dat)
        if(dat['closeDate'] >= "2022-01-01" and dat['closeDate'] <= "2022-12-31"):
            data_22.append(dat)
        if(dat['closeDate'] >= "2023-01-01" and dat['closeDate'] <= "2023-12-31"):
            data_23.append(dat)
    c=c+1
    print(round(c/3006*100,2), sep=", ", end="% ")

0.03% 0.07% 0.1% 0.13% 0.17% 0.2% 0.23% 0.27% 0.3% 0.33% 0.37% 0.4% 0.43% 0.47% 0.5% 0.53% 0.57% 0.6% 0.63% 0.67% 0.7% 0.73% 0.77% 0.8% 0.83% 0.86% 0.9% 0.93% 0.96% 1.0% 1.03% 1.06% 1.1% 1.13% 1.16% 1.2% 1.23% 1.26% 1.3% 1.33% 1.36% 1.4% 1.43% 1.46% 1.5% 1.53% 1.56% 1.6% 1.63% 1.66% 1.7% 1.73% 1.76% 1.8% 1.83% 1.86% 1.9% 1.93% 1.96% 2.0% 2.03% 2.06% 2.1% 2.13% 2.16% 2.2% 2.23% 2.26% 2.3% 2.33% 2.36% 2.4% 2.43% 2.46% 2.5% 2.53% 2.56% 2.59% 2.63% 2.66% 2.69% 2.73% 2.76% 2.79% 2.83% 2.86% 2.89% 2.93% 2.96% 2.99% 3.03% 3.06% 3.09% 3.13% 3.16% 3.19% 3.23% 3.26% 3.29% 3.33% 3.36% 3.39% 3.43% 3.46% 3.49% 3.53% 3.56% 3.59% 3.63% 3.66% 3.69% 3.73% 3.76% 3.79% 3.83% 3.86% 3.89% 3.93% 3.96% 3.99% 4.03% 4.06% 4.09% 4.13% 4.16% 4.19% 4.22% 4.26% 4.29% 4.32% 4.36% 4.39% 4.42% 4.46% 4.49% 4.52% 4.56% 4.59% 4.62% 4.66% 4.69% 4.72% 4.76% 4.79% 4.82% 4.86% 4.89% 4.92% 4.96% 4.99% 5.02% 5.06% 5.09% 5.12% 5.16% 5.19% 5.22% 5.26% 5.29% 5.32% 5.36% 5.39% 5.42% 5.46% 5.49% 5.52% 5.56% 5.59% 5.62% 5.66% 5.69%

<b>Exporting all datas in json as back to avoid redundant API calls</b>

In [72]:
df_data_20_1 = pd.DataFrame(data_20_1)                                                  
df_data_20_1.to_json('data_20_1.json', orient='records', indent=4)
df_data_21_1 = pd.DataFrame(data_21_1)
df_data_21_1.to_json('data_21_1.json', orient='records', indent=4)
df_data_22_1 = pd.DataFrame(data_22_1)
df_data_22_1.to_json('data_22_1.json', orient='records', indent=4)
df_data_23_1 = pd.DataFrame(data_23_1)
df_data_23_1.to_json('data_23_1.json', orient='records', indent=4)

Creating different methods that collects details of different reviews within different time frame. Ex - <b>Past Quarter, Since 2021, Since 2020, etc.</b>

In [ ]:
def previous_3_months_time():
    time = {}
    rev = {}
    df = pd.DataFrame()
    for i in data_23:
        if(i['closeDate'] > "2023-09-31" and i['closeDate'] <= "2023-12-31"):
            for j in i['reviewers']['reviewer']:
                if "timeSpent" in j:
                    time[j['userName']] = 0
                    rev[j['userName']] = 0
        
    for i in data_23:
        if(i['closeDate'] > "2023-09-31" and i['closeDate'] <= "2023-12-31"):
            for j in i['reviewers']['reviewer']:
                if "timeSpent" in j:
                    time[j['userName']] = time[j['userName']] + round(j['timeSpent']/3600000,2)
                    rev[j['userName']] = rev[j['userName']] + 1
        
    df_t = pd.DataFrame(list(time.items()), columns=['user', 'timeSpent'])
    df_r = pd.DataFrame(list(rev.items()), columns=['user', 'No. of reviews'])
    df_t = pd.merge(df_t, df_r, on='user')
    
    for j in data_23:
        if(j['closeDate'] > "2023-09-31" and j['closeDate'] <= "2023-12-31"):
            for i in j['stats']:
                df = df.append(i, ignore_index=True)
            
    df = df.drop(columns=['drafts', 'unread', 'leaveUnread', 'read'])
    df = df.groupby('user').sum().reset_index()
    df = pd.merge(df, df_t, on='user')
    
    global def_qtr
    def_qtr = df
    def_qtr = def_qtr.rename(columns={"user" : "Asst. ID", "published" : "Comments", "defects" : "Defects", "timeSpent" : "Time spent (hrs)"})
    def_qtr = def_qtr.sort_values(by='Defects', ascending=False)
    return df

def previous_6_months_time():
    previous_3_months_time()
    time = {}
    rev = {}
    df = pd.DataFrame()
    for i in data_23:
        if(i['closeDate'] > "2023-06-30" and i['closeDate'] <= "2023-12-31"):
            for j in i['reviewers']['reviewer']:
                if "timeSpent" in j:
                    time[j['userName']] = 0
                    rev[j['userName']] = 0
        
    for i in data_23:
        if(i['closeDate'] > "2023-06-30" and i['closeDate'] <= "2023-12-31"):
            for j in i['reviewers']['reviewer']:
                if "timeSpent" in j:
                    time[j['userName']] = time[j['userName']] + round(j['timeSpent']/3600000,2)
                    rev[j['userName']] = rev[j['userName']] + 1
        
    df_t = pd.DataFrame(list(time.items()), columns=['user', 'timeSpent'])
    df_r = pd.DataFrame(list(rev.items()), columns=['user', 'No. of reviews'])
    df_t = pd.merge(df_t, df_r, on='user')
    
    for j in data_23:
        if(j['closeDate'] > "2023-06-30" and j['closeDate'] <= "2023-12-31"):
            for i in j['stats']:
                df = df.append(i, ignore_index=True)
            
    df = df.drop(columns=['drafts', 'unread', 'leaveUnread', 'read'])
    df = df.groupby('user').sum().reset_index()
    df = pd.merge(df, df_t, on='user')
    
    global def_half
    def_half = df
    def_half = def_half.rename(columns={"user" : "Asst. ID", "published" : "Comments", "defects" : "Defects", "timeSpent" : "Time spent (hrs)"})
    def_half = def_half.sort_values(by='Defects', ascending=False)
    return df

def previous_year_time():
    previous_6_months_time()
    time = {}
    rev = {}
    df = pd.DataFrame()
    for i in data_23:
        for j in i['reviewers']['reviewer']:
            if "timeSpent" in j:
                time[j['userName']] = 0
                rev[j['userName']] = 0
        
    for i in data_23:
        for j in i['reviewers']['reviewer']:
            if "timeSpent" in j:
                time[j['userName']] = time[j['userName']] + round(j['timeSpent']/3600000,2)
                rev[j['userName']] = rev[j['userName']] + 1
        
    df_t = pd.DataFrame(list(time.items()), columns=['user', 'timeSpent'])
    df_r = pd.DataFrame(list(rev.items()), columns=['user', 'No. of reviews'])
    df_t = pd.merge(df_t, df_r, on='user')
    
    # For calculating Number of comments and defects
    for j in data_23:
        for i in j['stats']:
            df = df.append(i, ignore_index=True)
            
    df = df.drop(columns=['drafts', 'unread', 'leaveUnread', 'read'])
    df = df.groupby('user').sum().reset_index()
    df = pd.merge(df, df_t, on='user')
    
    global def_23
    def_23 = df
    def_23 = def_23.rename(columns={"user" : "Asst. ID", "published" : "Comments", "defects" : "Defects", "timeSpent" : "Time spent (hrs)"})
    def_23 = def_23.sort_values(by='Defects', ascending=False)
    return df

def previous_two_year_time():
    time = {}
    rev = {}
    df = pd.DataFrame()
    df_yr = previous_year_time()
    for i in data_22:
        for j in i['reviewers']['reviewer']:
            if "timeSpent" in j:
                time[j['userName']] = 0
                rev[j['userName']] = 0
        
    for i in data_22:
        for j in i['reviewers']['reviewer']:
            if "timeSpent" in j:
                time[j['userName']] = time[j['userName']] + round(j['timeSpent']/3600000,2)
                rev[j['userName']] = rev[j['userName']] + 1
    
    
    df_t = pd.DataFrame(list(time.items()), columns=['user', 'timeSpent'])
    df_r = pd.DataFrame(list(rev.items()), columns=['user', 'No. of reviews'])
    df_t = pd.merge(df_t, df_r, on='user')
    
    merged_df = pd.merge(df_t, df_yr, on='user', how='outer')
    merged_df.fillna(0.0, inplace = True)
    merged_df["timeSpent"] = merged_df["timeSpent_x"] + merged_df["timeSpent_y"]
    merged_df.drop(columns=["timeSpent_x", "timeSpent_y"], inplace = True)
    
    # For calculating Number of comments and defects
    for j in data_22:
        for i in j['stats']:
            df = df.append(i, ignore_index=True)
            
    df = df.drop(columns=['drafts', 'unread', 'leaveUnread', 'read'])
    df = df.groupby('user').sum().reset_index()
    df = pd.merge(df, merged_df, on='user')
    df.fillna(0.0, inplace = True)
    df["published"] = df["published_x"] + df["published_y"]
    df["defects"] = df["defects_x"] + df["defects_y"]
    df["No. of reviews"] = df["No. of reviews_x"] + df["No. of reviews_y"]
    df.drop(columns=["published_x", "published_y", "defects_x", "defects_y", "No. of reviews_x", "No. of reviews_y"], inplace = True)
    
    global def_22
    def_22 = df
    def_22 = def_22.rename(columns={"user" : "Asst. ID", "published" : "Comments", "defects" : "Defects", "timeSpent" : "Time spent (hrs)"})
    def_22 = def_22.sort_values(by='Defects', ascending=False)
    return df

def previous_three_year_time():
    time = {}
    rev = {}
    df = pd.DataFrame()
    df_yr = previous_two_year_time()
    for i in data_21:
        for j in i['reviewers']['reviewer']:
            if "timeSpent" in j:
                time[j['userName']] = 0
                rev[j['userName']] = 0
        
    for i in data_21:
        for j in i['reviewers']['reviewer']:
            if "timeSpent" in j:
                time[j['userName']] = time[j['userName']] + round(j['timeSpent']/3600000,2)
                rev[j['userName']] = rev[j['userName']] + 1
        
    df_t = pd.DataFrame(list(time.items()), columns=['user', 'timeSpent'])
    df_r = pd.DataFrame(list(rev.items()), columns=['user', 'No. of reviews'])
    df_t = pd.merge(df_t, df_r, on='user')
    merged_df = pd.merge(df_t, df_yr, on='user', how='outer')
    merged_df.fillna(0.0, inplace = True)
    merged_df["timeSpent"] = merged_df["timeSpent_x"] + merged_df["timeSpent_y"]
    merged_df.drop(columns=["timeSpent_x", "timeSpent_y"], inplace = True)
    
    # For calculating Number of comments and defects
    for j in data_21:
        for i in j['stats']:
            df = df.append(i, ignore_index=True)
            
    df = df.drop(columns=['drafts', 'unread', 'leaveUnread', 'read'])
    df = df.groupby('user').sum().reset_index()
    df = pd.merge(df, merged_df, on='user')
    df.fillna(0.0, inplace = True)
    df["published"] = df["published_x"] + df["published_y"]
    df["defects"] = df["defects_x"] + df["defects_y"]
    df["No. of reviews"] = df["No. of reviews_x"] + df["No. of reviews_y"]
    df.drop(columns=["published_x", "published_y", "defects_x", "defects_y", "No. of reviews_x", "No. of reviews_y"], inplace = True)
    
    global def_21
    def_21 = df
    def_21 = def_21.rename(columns={"user" : "Asst. ID", "published" : "Comments", "defects" : "Defects", "timeSpent" : "Time spent (hrs)"})
    def_21 = def_21.sort_values(by='Defects', ascending=False)
    return df

def previous_four_year_time():
    time = {}
    rev = {}
    df = pd.DataFrame()
    df_yr = previous_three_year_time()
    for i in data_20:
        for j in i['reviewers']['reviewer']:
            if "timeSpent" in j:
                time[j['userName']] = 0
                rev[j['userName']] = 0
        
    for i in data_20:
        for j in i['reviewers']['reviewer']:
            if "timeSpent" in j:
                time[j['userName']] = time[j['userName']] + round(j['timeSpent']/3600000,2)
                rev[j['userName']] = rev[j['userName']] + 1
        
    df_t = pd.DataFrame(list(time.items()), columns=['user', 'timeSpent'])
    df_r = pd.DataFrame(list(rev.items()), columns=['user', 'No. of reviews'])
    df_t = pd.merge(df_t, df_r, on='user')
    merged_df = pd.merge(df_t, df_yr, on='user')
    merged_df.fillna(0.0, inplace = True)
    merged_df["timeSpent"] = merged_df["timeSpent_x"] + merged_df["timeSpent_y"]
    merged_df.drop(columns=["timeSpent_x", "timeSpent_y"], inplace = True)
    
    # For calculating Number of comments and defects
    for j in data_20:
        for i in j['stats']:
            df = df.append(i, ignore_index=True)
            
    df = df.drop(columns=['drafts', 'unread', 'leaveUnread', 'read'])
    df = df.groupby('user').sum().reset_index()
    df = pd.merge(df, merged_df, on='user')
    df.fillna(0.0, inplace = True)
    df["published"] = df["published_x"] + df["published_y"]
    df["defects"] = df["defects_x"] + df["defects_y"]
    df["No. of reviews"] = df["No. of reviews_x"] + df["No. of reviews_y"]
    df.drop(columns=["published_x", "published_y", "defects_x", "defects_y", "No. of reviews_x", "No. of reviews_y"], inplace = True)
    
    global def_20
    def_20 = df
    def_20 = def_20.rename(columns={"user" : "Asst. ID", "published" : "Comments", "defects" : "Defects", "timeSpent" : "Time spent (hrs)"})
    def_20 = def_20.sort_values(by='Defects', ascending=False)
    return df

df = previous_four_year_time()
df.head()

#### Converting all associate ID into uppercase for searching Convince

In [ ]:
def_qtr["Asst. ID"] = def_qtr["Asst. ID"].str.upper()      
def_half["Asst. ID"] = def_half["Asst. ID"].str.upper()
def_23["Asst. ID"] = def_23["Asst. ID"].str.upper()
def_22["Asst. ID"] = def_22["Asst. ID"].str.upper()
def_21["Asst. ID"] = def_21["Asst. ID"].str.upper()
def_20["Asst. ID"] = def_20["Asst. ID"].str.upper()
def_qtr.head()

#### Exporting dataframe into csv

In [246]:
def_qtr.to_csv('Last Quarter.csv', index=False)
def_half.to_csv('Last 6 months.csv', index=False)
def_23.to_csv('Since 23.csv', index=False)
def_22.to_csv('Since 22.csv', index=False)
def_21.to_csv('Since 21.csv', index=False)
def_20.to_csv('Since 20.csv', index=False)

## GUI

In [296]:
import gradio as gr
import plotly.express as px

In [295]:
%%time
def tables(yr):
    if(yr=='Last Quater'):
        def_qtr["Time spent (hrs)"] = def_qtr["Time spent (hrs)"].round(2)
        return def_qtr
    if(yr=='Past 6 months'):
        def_half["Time spent (hrs)"] = def_half["Time spent (hrs)"].round(2)
        return def_half            
    if(yr=='Since 2023'):
        def_23["Time spent (hrs)"] = def_23["Time spent (hrs)"].round(2)
        return def_23            
    if(yr=='Since 2022'):
        def_22["Time spent (hrs)"] = def_22["Time spent (hrs)"].round(2)
        return def_22            
    if(yr=='Since 2021'):
        def_21["Time spent (hrs)"] = def_21["Time spent (hrs)"].round(2)
        return def_21            
    if(yr=='Since 2020'):
        def_20["Time spent (hrs)"] = def_20["Time spent (hrs)"].round(2)
        return def_20
    
    
def associate(yr, aid=None):
    df = tables(yr)
    if(aid==None):
        return df
    else:
        return df[df["Asst. ID"] == aid]
    

def Plots(yr, stat):
    if(yr=='Last Quater'):
        if(stat=='Defects'):
            def_qtr.sort_values(by='Defects', ascending=False, inplace = True)
            fig = px.bar(def_qtr, x='Asst. ID', y=def_qtr["Defects"])
            fig.update_layout(title="Defects", yaxis_title="No. of Defects", xaxis_title="Associate")
        if(stat=='Comments'):
            def_qtr.sort_values(by='Comments', ascending=False, inplace = True)
            fig = px.bar(def_qtr, x='Asst. ID', y=def_qtr["Comments"])
            fig.update_layout(title="Comments", yaxis_title="No. of Defects", xaxis_title="Associate")
        if(stat=='Time Spent'):
            def_qtr.sort_values(by='Time spent (hrs)', ascending=False, inplace = True)
            fig = px.bar(def_qtr, x='Asst. ID', y=def_qtr["Time spent (hrs)"])
            fig.update_layout(title="Time Spent", yaxis_title="Hours", xaxis_title="Associate")
        if(stat=='No. of reviews'):
            def_qtr.sort_values(by='No. of reviews', ascending=False, inplace = True)
            fig = px.bar(def_qtr, x='Asst. ID', y=def_qtr["No. of reviews"])
            fig.update_layout(title="No. of reviews", yaxis_title="Reviews", xaxis_title="Associate")
            
    if(yr=='Past 6 months'):
        if(stat=='Defects'):
            def_half.sort_values(by='Defects', ascending=False, inplace = True)
            fig = px.bar(def_half, x='Asst. ID', y=def_half["Defects"])
            fig.update_layout(title="Defects", yaxis_title="No. of Defects", xaxis_title="Associate")
        if(stat=='Comments'):
            def_half.sort_values(by='Comments', ascending=False, inplace = True)
            fig = px.bar(def_half, x='Asst. ID', y=def_half["Comments"])
            fig.update_layout(title="Comments", yaxis_title="No. of Defects", xaxis_title="Associate")
        if(stat=='Time Spent'):
            def_half.sort_values(by='Time spent (hrs)', ascending=False, inplace = True)
            fig = px.bar(def_half, x='Asst. ID', y=def_half["Time spent (hrs)"])
            fig.update_layout(title="Time Spent", yaxis_title="Hours", xaxis_title="Associate")
        if(stat=='No. of reviews'):
            def_half.sort_values(by='No. of reviews', ascending=False, inplace = True)
            fig = px.bar(def_half, x='Asst. ID', y=def_half["No. of reviews"])
            fig.update_layout(title="No. of reviews", yaxis_title="Reviews", xaxis_title="Associate")
            
    if(yr=='Since 2023'):
        if(stat=='Defects'):
            def_23.sort_values(by='Defects', ascending=False, inplace = True)
            fig = px.bar(def_23, x='Asst. ID', y=def_23["Defects"])
            fig.update_layout(title="Defects", yaxis_title="No. of Defects", xaxis_title="Associate")
        if(stat=='Comments'):
            def_23.sort_values(by='Comments', ascending=False, inplace = True)
            fig = px.bar(def_23, x='Asst. ID', y=def_23["Comments"])
            fig.update_layout(title="Comments", yaxis_title="No. of Defects", xaxis_title="Associate")
        if(stat=='Time Spent'):
            def_23.sort_values(by='Time spent (hrs)', ascending=False, inplace = True)
            fig = px.bar(def_23, x='Asst. ID', y=def_23["Time spent (hrs)"])
            fig.update_layout(title="Time Spent", yaxis_title="Hours", xaxis_title="Associate")
        if(stat=='No. of reviews'):
            def_23.sort_values(by='No. of reviews', ascending=False, inplace = True)
            fig = px.bar(def_23, x='Asst. ID', y=def_23["No. of reviews"])
            fig.update_layout(title="No. of reviews", yaxis_title="Reviews", xaxis_title="Associate")
            
    if(yr=='Since 2022'):
        if(stat=='Defects'):
            def_22.sort_values(by='Defects', ascending=False, inplace = True)
            fig = px.bar(def_22, x='Asst. ID', y=def_22["Defects"])
            fig.update_layout(title="Defects", yaxis_title="No. of Defects", xaxis_title="Associate")
        if(stat=='Comments'):
            def_22.sort_values(by='Comments', ascending=False, inplace = True)
            fig = px.bar(def_22, x='Asst. ID', y=def_22["Comments"])
            fig.update_layout(title="Comments", yaxis_title="No. of Defects", xaxis_title="Associate")
        if(stat=='Time Spent'):
            def_22.sort_values(by='Time spent (hrs)', ascending=False, inplace = True)
            fig = px.bar(def_22, x='Asst. ID', y=def_22["Time spent (hrs)"])
            fig.update_layout(title="Time Spent", yaxis_title="Hours", xaxis_title="Associate")
        if(stat=='No. of reviews'):
            def_22.sort_values(by='No. of reviews', ascending=False, inplace = True)
            fig = px.bar(def_22, x='Asst. ID', y=def_22["No. of reviews"])
            fig.update_layout(title="No. of reviews", yaxis_title="Reviews", xaxis_title="Associate")
            
    if(yr=='Since 2021'):
        if(stat=='Defects'):
            def_21.sort_values(by='Defects', ascending=False, inplace = True)
            fig = px.bar(def_21, x='Asst. ID', y=def_21["Defects"])
            fig.update_layout(title="Defects", yaxis_title="No. of Defects", xaxis_title="Associate")
        if(stat=='Comments'):
            def_21.sort_values(by='Comments', ascending=False, inplace = True)
            fig = px.bar(def_21, x='Asst. ID', y=def_21["Comments"])
            fig.update_layout(title="Comments", yaxis_title="No. of Defects", xaxis_title="Associate")
        if(stat=='Time Spent'):
            def_21.sort_values(by='Time spent (hrs)', ascending=False, inplace = True)
            fig = px.bar(def_21, x='Asst. ID', y=def_21["Time spent (hrs)"])
            fig.update_layout(title="Time Spent", yaxis_title="Hours", xaxis_title="Associate")
        if(stat=='No. of reviews'):
            def_21.sort_values(by='No. of reviews', ascending=False, inplace = True)
            fig = px.bar(def_21, x='Asst. ID', y=def_21["No. of reviews"])
            fig.update_layout(title="No. of reviews", yaxis_title="Reviews", xaxis_title="Associate")
            
    if(yr=='Since 2020'):
        if(stat=='Defects'):
            def_20.sort_values(by='Defects', ascending=False, inplace = True)
            fig = px.bar(def_20, x='Asst. ID', y=def_20["Defects"])
            fig.update_layout(title="Defects", yaxis_title="No. of Defects", xaxis_title="Associate")
        if(stat=='Comments'):
            def_20.sort_values(by='Comments', ascending=False, inplace = True)
            fig = px.bar(def_20, x='Asst. ID', y=def_20["Comments"])
            fig.update_layout(title="Comments", yaxis_title="No. of Defects", xaxis_title="Associate")
        if(stat=='Time Spent'):
            def_20.sort_values(by='Time spent (hrs)', ascending=False, inplace = True)
            fig = px.bar(def_20, x='Asst. ID', y=def_20["Time spent (hrs)"])
            fig.update_layout(title="Time Spent", yaxis_title="Hours", xaxis_title="Associate")
        if(stat=='No. of reviews'):
            def_20.sort_values(by='No. of reviews', ascending=False, inplace = True)
            fig = px.bar(def_20, x='Asst. ID', y=def_20["No. of reviews"])
            fig.update_layout(title="No. of reviews", yaxis_title="Reviews", xaxis_title="Associate")
    
    return fig


with gr.Blocks() as Tabular:
    inputs = [gr.Dropdown(["Since 2020", "Since 2021", "Since 2022", "Since 2023", "Past 6 months", "Last Quarter"], 
                         label="Year"),
             gr.Textbox(label="Associate ID", placeholder="(in UPPERCASE)")]
    out = gr.DataFrame()
    greet_btn = gr.Button("Search")
    inputs[0].change(tables, inputs[0], out)
    greet_btn.click(fn=associate, inputs=inputs, outputs=out, api_name="search")
    
with gr.Blocks() as Graphical:
    inputs = [gr.Dropdown(["Since 2020", "Since 2021", "Since 2022", "Since 2023", "Past 6 months", "Last Quarter"], 
                          label="Year"),
              gr.Radio(["Defects", "Comments", "Time Spent", "No. of reviews"], label="Review statictics")]
    out = gr.Plot()
    inputs[0].change(Plots, inputs, out)
    inputs[1].change(Plots, inputs, out)
    
demo = gr.TabbedInterface([Tabular, Graphical], ["Tabular", "Graphical"])

demo.close()
demo.launch()

Running on local URL:  http://127.0.0.1:7892

To create a public link, set `share=True` in `launch()`.


CPU times: user 625 ms, sys: 101 ms, total: 725 ms
Wall time: 1.84 s


# You can test the demo app in readme.md file